editor/inserter/index.js

```diff
@@ -1,18 +1,31 @@
+/**
+ * External dependencies
+ */
+import { groupBy } from 'lodash';
+
 const Inserter = () => {
 	const blocks = wp.blocks.getBlocks();
+	const blocksByCategory = groupBy( blocks, 'category' );
+	const categories = wp.blocks.getCategories();
 
 	return (
 		<div className="inserter">
 			<div className="inserter__arrow" />
 			<div className="inserter__content">
-				<div className="inserter__Category-blocks">
-					{ blocks.map( ( { slug, title, icon } ) => (
-						<div key={ slug } className="inserter__block">
-							<span className={ 'dashicons dashicons-' + icon } />
-							{ title }
+				{ categories
```

Comment:
You would probably just want to do `reduce` and `pipe` if you broke each mapping filtering function into something that could be called.  That way you loop once and apply the same chain sequentially.  If that makes sense.